lib/screens/v2/disruption_diagram/model.ex

```diff
@@ -1,81 +1,172 @@
-          # First and last elements of the list are `end_slot`s, middle elements are `middle_slot`s.
-          slots: list(slot())
-        }
-
-  @type discrete_disruption_diagram :: %{
-          effect: :station_closure,
-          closed_station_slot_indices: list(non_neg_integer()),
-          line: line_color(),
-          current_station_slot_index: non_neg_integer(),
-          # First and last elements of the list are `end_slot`s, middle elements are `middle_slot`s.
-          slots: list(slot())
-        }
-
-  @type slot :: end_slot() | middle_slot()
-
-  @type end_slot :: %{
-          type: :arrow | :terminal,
-          label_id: end_label_id()
-        }
-
-  @type middle_slot :: %{
-          label: label(),
-          show_symbol: boolean()
-        }
-
-  @type label :: ellipsis() | %{full: String.t(), abbrev: String.t()}
-
-  # Literally the string "…", but you can't use string literals as types in elixir
-  @type ellipsis :: String.t()
-
-  # End labels have hardcoded presentation, so we just send an ID for the client to use in
-  # a lookup.
-  #
-  # TBD what these IDs will look like. We might just use parent station IDs.
-  #
-  # The rest of the labels' presentations are computed based on the height of the end labels,
-  # so we can send actual text for those--it will be dynamically resized to fit.
-  @type end_label_id :: String.t()
-
-  @type line_color :: :blue | :orange | :red | :green
+  alias Screens.V2.DisruptionDiagram, as: DD
+  alias Screens.V2.DisruptionDiagram.Builder, as: B
+  alias Screens.V2.DisruptionDiagram.Validator
+  alias Screens.V2.LocalizedAlert
+
+  import LocalizedAlert, only: [is_localized_alert: 1]
+
+  # If the diagram is shorter than 6 slots, we "pad" it until it contains at least 6.
+  @minimum_slot_count 6
+
+  # If the closure is longer than 8 stops, it needs to be collapsed.
+  @max_closure_count 8
+
+  # When the closure needs to be collapsed, we omit stops
+  # from it until the diagram contains 12 slots total.
+  @max_count_with_collapsed_closure 12
+
+  # When the closure needs to be collapsed, we automatically
+  # also collapse the gap, making it take 2 slots or fewer.
+  @max_collapsed_gap_count 2
+
+  # If everything else fits, we still limit the gap to 3 slots or fewer.
+  @max_gap_count 3
 
   @doc "Produces a JSON-serializable map representing the disruption diagram."
-  # Update spec when this gets implemented!
-  @spec serialize(t()) :: nil
-  def serialize(_model) do
-    nil
+  @spec serialize(DD.t()) :: {:ok, DD.serialized_response()} | {:error, reason :: String.t()}
+  def serialize(localized_alert) when is_localized_alert(localized_alert) do
+    with :ok <- Validator.validate(localized_alert) do
+      do_serialize(localized_alert)
+    end
+  rescue
+    error ->
+      error_string =
+        Exception.message(error) <> "\n\n" <> Exception.format_stacktrace(__STACKTRACE__)
+
+      {:error, "Exception raised during serialization:\n\n#{error_string}"}
```

Comment:
Sorry, this turned into a novel...

<details>
<summary><h3><code>rescue</code> / implicit try</h2></summary>

---
`rescue` is basically the `catch` clause of a try/catch. For some reason in elixir they call it `try`/`rescue` instead, but it works exactly the same—it catches any raised exceptions. If the try/catch is at the top level of the function, you can make it "implicit", which is what I did here. ([docs](https://hexdocs.pm/elixir/Kernel.html#def/2-rescue-catch-after-else))
If this code feels too "clever" I'd be happy to make the `try` expression explicit instead!

---

</details>

<details>
<summary><h3><code>with</code> without <code>else</code></h3></summary>

---

When a `with` doesn't have an `else` block, it just immediately returns anything that doesn't match one of the arrow clauses. Unlike the `->`'s in a `case` expression, a left-arrow clause `<-` doesn't raise an exception when a match fails. This is true in `for` expressions as well!

This:
```ex
with {:ok, data} <- fetch_data(),
     {:ok, entity} <- parse_data(data),
     {:ok, related_entity} <- Map.fetch(relation, entity.id) do
  {:ok, entity, related_entity}
end
```
is equivalent to this:
```ex
with {:ok, data} <- fetch_data(),
     {:ok, entity} <- parse_data(data),
     {:ok, related_entity} <- Map.fetch(relation, entity.id) do
  {:ok, entity, related_entity}
else
  non_matching_value -> non_matching_value
end
```
The first function call that returns something other than `{:ok, _some_value}` will make the `with` immediately return that value.
For example, if `fetch_data` succeeds but `parse_data` returns `{:error, some_message}` then the whole `with` will return `{:error, some_message}`.

You only need an `else` block if you need to modify some non-happy-path values to something else before returning. ([docs](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1-else-clauses))

If all of the called functions return something of the same shape for their failure cases, e.g. `{:error, message}`, then you don't really need an `else` block in your `with` expression. You can just return the error value as-is.

---

</details>

So basically, I was able to avoid `else` clauses on my `with`s because all of the "graceful failure" cases return the same `{:error, message}` shape which falls through straight to the top. That case is handled [here](https://github.com/mbta/screens/pull/1828/files#diff-14e0fcdf7283a1020ce87a425d3445b064821f56e9ee5034235bd3d5d7d73e4fR892) and documented in the [spec](https://github.com/mbta/screens/blob/d842bf7d8a7056bfdae16dfba771b34198cdc098/lib/screens/v2/disruption_diagram/model.ex#L31).

The `rescue` is there to separately deal with any _unexpected_ failures that result in an exception being raised, e.g. failing to look up a stop name [here](https://github.com/mbta/screens/blob/d842bf7d8a7056bfdae16dfba771b34198cdc098/lib/screens/v2/disruption_diagram/builder.ex#L117).

I wanted to make sure that **no matter what, the disruption diagram code can't cause the whole request to fail**, since it's an optional "nice-to-have" part of the alert widget.

Reply:
Oh my! Waaaaay more detail than expected, but definitely thoughtful and makes all the sense. Thanks for the resources where I can learn more if needed. But sounds good to me